instrumentation/micrometer/micrometer-1.5/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/micrometer/v1_5/AsyncInstrumentRegistry.java

```diff
@@ -36,83 +44,230 @@
-                    .gaugeBuilder(meterId.getName())
+                    .gaugeBuilder(name)
                     .setDescription(description(meterId))
                     .setUnit(baseUnit(meterId))
                     .buildWithCallback(recorderCallback);
                 return recorderCallback;
               });
-      recorder.addGaugeMeasurement(attributes, obj, objMetric);
+      recorder.addMeasurement(
+          attributes, new DoubleMetricInfo(obj, (ToDoubleFunction<Object>) objMetric));
     }
   }
 
   void removeGauge(String name, Attributes attributes) {
     synchronized (gauges) {
-      GaugeMeasurementsRecorder recorder = gauges.get(name);
-      if (recorder != null) {
-        recorder.removeGaugeMeasurement(attributes);
-        // if this was the last measurement then let's remove the whole recorder
-        if (recorder.isEmpty()) {
-          gauges.remove(name);
-        }
-      }
+      removeMeasurement(gauges, name, attributes);
     }
   }
 
-  private final class GaugeMeasurementsRecorder implements Consumer<ObservableDoubleMeasurement> {
+  <T> void buildDoubleCounter(
+      io.micrometer.core.instrument.Meter.Id meterId,
+      Attributes attributes,
+      T obj,
+      ToDoubleFunction<T> objMetric) {
+    buildDoubleCounter(
+        meterId.getName(), description(meterId), baseUnit(meterId), attributes, obj, objMetric);
+  }
 
-    @GuardedBy("gauges")
-    private final Map<Attributes, GaugeInfo> measurements = new HashMap<>();
+  <T> void buildDoubleCounter(
+      String name,
+      String description,
+      String baseUnit,
+      Attributes attributes,
+      @Nullable T obj,
+      ToDoubleFunction<T> objMetric) {
 
-    @Override
-    public void accept(ObservableDoubleMeasurement measurement) {
-      Map<Attributes, GaugeInfo> measurementsCopy;
-      synchronized (gauges) {
-        measurementsCopy = new HashMap<>(measurements);
+    synchronized (doubleCounters) {
+      // use the counters map as lock for the recorder state - this way all double counter-related
+      // mutable state will always be accessed in synchronized(doubleCounters)
+      Object recorderLock = doubleCounters;
+
+      DoubleMeasurementsRecorder recorder =
+          doubleCounters.computeIfAbsent(
+              name,
+              n -> {
+                DoubleMeasurementsRecorder recorderCallback =
+                    new DoubleMeasurementsRecorder(recorderLock);
+                meter
+                    .counterBuilder(name)
+                    .setDescription(description)
+                    .setUnit(baseUnit)
+                    .ofDoubles()
+                    .buildWithCallback(recorderCallback);
+                return recorderCallback;
+              });
+      recorder.addMeasurement(
+          attributes, new DoubleMetricInfo(obj, (ToDoubleFunction<Object>) objMetric));
+    }
+  }
+
+  void removeDoubleCounter(String name, Attributes attributes) {
+    synchronized (doubleCounters) {
```

Comment:
#nit: can move the synchronization to inside `removeMeasurement` to DRY up the code a bit.

Reply:
Unfortunately errorprone doesn't allow to compile it (because of `@GuardedBy`), so I have to duplicate this `synchronized` block.